// ngx-file-helpers/esm5/lib/file-dropzone.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import { FileHandler } from './file-handler';
var FileDropzoneDirective = /** @class */ (function (_super) {
    __extends(FileDropzoneDirective, _super);
    function FileDropzoneDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.fileDrop = new EventEmitter();
        return _this;
    }
    FileDropzoneDirective.prototype.onDragEnter = function (event) {
        event.stopPropagation();
        event.preventDefault();
    };
    FileDropzoneDirective.prototype.onDragOver = function (event) {
        event.stopPropagation();
        event.preventDefault();
    };
    FileDropzoneDirective.prototype.onDrop = function (event) {
        var _this = this;
        event.stopPropagation();
        event.preventDefault();
        this.readFiles(event.dataTransfer.files, function (readFile) {
            return _this.fileDrop.emit(readFile);
        });
    };
    __decorate([
        Output()
    ], FileDropzoneDirective.prototype, "fileDrop", void 0);
    __decorate([
        HostListener('dragenter', ['$event'])
    ], FileDropzoneDirective.prototype, "onDragEnter", null);
    __decorate([
        HostListener('dragover', ['$event'])
    ], FileDropzoneDirective.prototype, "onDragOver", null);
    __decorate([
        HostListener('drop', ['$event'])
    ], FileDropzoneDirective.prototype, "onDrop", null);
    FileDropzoneDirective = __decorate([
        Directive({
            selector: '[ngxFileDropzone]',
            exportAs: 'ngxFileDropzone'
        })
    ], FileDropzoneDirective);
    return FileDropzoneDirective;
}(FileHandler));
export { FileDropzoneDirective };
