// ngx-file-helpers/esm5/lib/file-picker.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, ElementRef, EventEmitter, HostListener, Input, OnInit, Output, Renderer2 } from '@angular/core';
import { FileHandler } from './file-handler';
import { coerceBooleanProperty } from './helpers';
var FilePickerDirective = /** @class */ (function (_super) {
    __extends(FilePickerDirective, _super);
    function FilePickerDirective(el, renderer) {
        var _this = _super.call(this) || this;
        _this.el = el;
        _this.renderer = renderer;
        _this.accept = '';
        _this.filePick = new EventEmitter();
        return _this;
    }
    Object.defineProperty(FilePickerDirective.prototype, "multiple", {
        get: function () {
            return this._multiple;
        },
        set: function (value) {
            this._multiple = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    FilePickerDirective.prototype.ngOnInit = function () {
        var _this = this;
        this._input = this.renderer.createElement('input');
        this.renderer.appendChild(this.el.nativeElement, this._input);
        this.renderer.setAttribute(this._input, 'type', 'file');
        this.renderer.setAttribute(this._input, 'accept', this.accept);
        this.renderer.setStyle(this._input, 'display', 'none');
        if (this.multiple) {
            this.renderer.setAttribute(this._input, 'multiple', 'multiple');
        }
        this.renderer.listen(this._input, 'change', function (event) {
            return _this._onListen(event);
        });
    };
    FilePickerDirective.prototype.reset = function () {
        if (!this._input) {
            console.error('It seems that ngOnInit() has not been executed or that the hidden _input element is null. Did you mess with the DOM?');
            return;
        }
        this._input.value = null;
    };
    FilePickerDirective.prototype.browse = function () {
        if (!this._input) {
            console.error('It seems that ngOnInit() has not been executed or that the hidden _input element is null. Did you mess with the DOM?');
            return;
        }
        this._input.click();
    };
    // The callback signature prevent the async/await usage
    FilePickerDirective.prototype._onListen = function (event) {
        var _this = this;
        var target = event.target;
        this.readFiles(target.files, function (readFile) { return _this.filePick.emit(readFile); })
            // reset value to trick change event making it changeable every time
            .finally(function () { return (target.value = ''); });
    };
    FilePickerDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input()
    ], FilePickerDirective.prototype, "accept", void 0);
    __decorate([
        Input()
    ], FilePickerDirective.prototype, "multiple", null);
    __decorate([
        Output()
    ], FilePickerDirective.prototype, "filePick", void 0);
    __decorate([
        HostListener('click')
    ], FilePickerDirective.prototype, "browse", null);
    FilePickerDirective = __decorate([
        Directive({
            selector: '[ngxFilePicker]',
            exportAs: 'ngxFilePicker'
        })
    ], FilePickerDirective);
    return FilePickerDirective;
}(FileHandler));
export { FilePickerDirective };
