// ngx-file-helpers/esm2015/lib/file-dropzone.directive.js
import { __decorate } from "tslib";
import { Directive, EventEmitter, HostListener, Output } from '@angular/core';
import { FileHandler } from './file-handler';
let FileDropzoneDirective = class FileDropzoneDirective extends FileHandler {
    constructor() {
        super(...arguments);
        this.fileDrop = new EventEmitter();
    }
    onDragEnter(event) {
        event.stopPropagation();
        event.preventDefault();
    }
    onDragOver(event) {
        event.stopPropagation();
        event.preventDefault();
    }
    onDrop(event) {
        event.stopPropagation();
        event.preventDefault();
        this.readFiles(event.dataTransfer.files, readFile => this.fileDrop.emit(readFile));
    }
};
__decorate([
    Output()
], FileDropzoneDirective.prototype, "fileDrop", void 0);
__decorate([
    HostListener('dragenter', ['$event'])
], FileDropzoneDirective.prototype, "onDragEnter", null);
__decorate([
    HostListener('dragover', ['$event'])
], FileDropzoneDirective.prototype, "onDragOver", null);
__decorate([
    HostListener('drop', ['$event'])
], FileDropzoneDirective.prototype, "onDrop", null);
FileDropzoneDirective = __decorate([
    Directive({
        selector: '[ngxFileDropzone]',
        exportAs: 'ngxFileDropzone'
    })
], FileDropzoneDirective);
export { FileDropzoneDirective };
