// ngx-file-helpers/esm2015/lib/file-picker.directive.js
import { __decorate } from "tslib";
import { Directive, ElementRef, EventEmitter, HostListener, Input, OnInit, Output, Renderer2 } from '@angular/core';
import { FileHandler } from './file-handler';
import { coerceBooleanProperty } from './helpers';
let FilePickerDirective = class FilePickerDirective extends FileHandler {
    constructor(el, renderer) {
        super();
        this.el = el;
        this.renderer = renderer;
        this.accept = '';
        this.filePick = new EventEmitter();
    }
    get multiple() {
        return this._multiple;
    }
    set multiple(value) {
        this._multiple = coerceBooleanProperty(value);
    }
    ngOnInit() {
        this._input = this.renderer.createElement('input');
        this.renderer.appendChild(this.el.nativeElement, this._input);
        this.renderer.setAttribute(this._input, 'type', 'file');
        this.renderer.setAttribute(this._input, 'accept', this.accept);
        this.renderer.setStyle(this._input, 'display', 'none');
        if (this.multiple) {
            this.renderer.setAttribute(this._input, 'multiple', 'multiple');
        }
        this.renderer.listen(this._input, 'change', (event) => this._onListen(event));
    }
    reset() {
        if (!this._input) {
            console.error('It seems that ngOnInit() has not been executed or that the hidden _input element is null. Did you mess with the DOM?');
            return;
        }
        this._input.value = null;
    }
    browse() {
        if (!this._input) {
            console.error('It seems that ngOnInit() has not been executed or that the hidden _input element is null. Did you mess with the DOM?');
            return;
        }
        this._input.click();
    }
    // The callback signature prevent the async/await usage
    _onListen(event) {
        const target = event.target;
        this.readFiles(target.files, readFile => this.filePick.emit(readFile))
            // reset value to trick change event making it changeable every time
            .finally(() => (target.value = ''));
    }
};
FilePickerDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    Input()
], FilePickerDirective.prototype, "accept", void 0);
__decorate([
    Input()
], FilePickerDirective.prototype, "multiple", null);
__decorate([
    Output()
], FilePickerDirective.prototype, "filePick", void 0);
__decorate([
    HostListener('click')
], FilePickerDirective.prototype, "browse", null);
FilePickerDirective = __decorate([
    Directive({
        selector: '[ngxFilePicker]',
        exportAs: 'ngxFilePicker'
    })
], FilePickerDirective);
export { FilePickerDirective };
